Disable fallback for custom kernels

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #101131

Previous failed attempt was here: https://github.com/pytorch/pytorch/pull/97715.
Basically we tried to disable fallback for all ops (aten + custom) but hit many CI failures due to missing fake tensor coverage. Let's just disable it for custom kernels for now. 


cc @ezyang @gchanan @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @desertfire

tugsbayasgalan committed:
Disable fallback for custom kernels

[ghstack-poisoned]

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1307,14 +1307,20 @@ def dispatch(self, func, types, args=(), kwargs=None):
+            if not self.allow_fallback_kernels:
+                return False
+            # should't fallback on custom ops
+            return func.namespace in ["prims", "aten"]
```
Also add "debugprims"

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1307,14 +1307,20 @@ def dispatch(self, func, types, args=(), kwargs=None):
+        def can_fallback(func: OpOverload):
+            if not self.allow_fallback_kernels:
+                return False
+            # should't fallback on custom ops
```
nit: leave a comment with some more context? Something like "It's OK to try the fallback for built-in ops (e.g. aten, prims) because we control and test these but the fallback leads to unexpected behavior in user-defined custom ops"

tugsbayasgalan committed:
Update on "Disable fallback for custom kernels"


Previous failed attempt was here: https://github.com/pytorch/pytorch/pull/97715.
Basically we tried to disable fallback for all ops (aten + custom) but hit many CI failures due to missing fake tensor coverage. Let's just disable it for custom kernels for now. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Disable fallback for custom kernels"


Previous failed attempt was here: https://github.com/pytorch/pytorch/pull/97715.
Basically we tried to disable fallback for all ops (aten + custom) but hit many CI failures due to missing fake tensor coverage. Let's just disable it for custom kernels for now. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Disable fallback for custom kernels"


Previous failed attempt was here: https://github.com/pytorch/pytorch/pull/97715.
Basically we tried to disable fallback for all ops (aten + custom) but hit many CI failures due to missing fake tensor coverage. Let's just disable it for custom kernels for now. 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "Disable fallback for custom kernels"


Previous failed attempt was here: https://github.com/pytorch/pytorch/pull/97715.
Basically we tried to disable fallback for all ops (aten + custom) but hit many CI failures due to missing fake tensor coverage. Let's just disable it for custom kernels for now. 


cc voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "Disable fallback for custom kernels"


Previous failed attempt was here: https://github.com/pytorch/pytorch/pull/97715.
Basically we tried to disable fallback for all ops (aten + custom) but hit many CI failures due to missing fake tensor coverage. Let's just disable it for custom kernels for now. 


cc voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented:
This PR disables the FakeTensor fallback for user-defined custom operators for (1) safety and (2) improve the error message when using a custom operator that doesn't support FakeTensor. If the custom operator does not have a meta kernel (or a FakeTensor kernel), the FakeTensor fallback attempts to re-use the backend (CPU/CUDA) kernel as the meta kernel and sends FakeTensor through to it. This led to several difficult-to-debug situations when using torch.compile with custom operators; e.g. it is sometimes possible to grab the (null) .data_ptr of a FakeTensor and access its memory.

This is a BC-breaking change in that there may exist custom operators that previously worked with the FakeTensor fallback but we have disabled the FakeTensor fallback for them. To get the old behavior back, you can either:
- If using the new CustomOp API: use `impl_abstract` on your CPU impl to define the meta/FakeTensor kernels.
- If using the TORCH_LIBRARY APIs: register a meta kernel (using your CPU implementation)
